### DISCUSSION TITLE
Keeping remote control app in sync

### DISCUSSION THREAD

**Discussion by willeke1234** [GrandOrgue/grandorgue#1171] [Help]
I'm experimenting with a remote control app that imports an ODF and extracts the buttons, images and labels on the panels. I'm having some trouble keeping the app and GrandOrgue in sync.

After loading a sample set, changing settings and starting MIDI recording, GrandOrgue sends MIDI events for all MIDI objects. Is it possible to trigger this from my app by sending a MIDI event?

**Comment by rousseldenis** [comment]
> I'm experimenting with a remote control app that imports an ODF and extracts the buttons, images and labels on the panels. I'm having some trouble keeping the app and GrandOrgue in sync.

That could be very interesting. Could you share a little bit more ?

> After loading a sample set, changing settings and starting MIDI recording, GrandOrgue sends MIDI events for all MIDI objects. Is it possible to trigger this from my app by sending a MIDI event?

@oleg68 Controlling GrandOrgue from custom MIDI events sounds great. Do you think it's possible ?

**Reply by oleg68** [reply]
> Controlling GrandOrgue from custom MIDI events sounds great. Do you think it's possible ?

Yes, it is possible with the current version of GrandOrgue. I need more information what features are missing.

**Comment by oleg68** [comment]
@willeke1234 
> After loading a sample set, changing settings and starting MIDI recording, GrandOrgue sends MIDI events for all MIDI objects. Is it possible to trigger this from my app by sending a MIDI event?

Could you explain more, what is wrong in the current GO behavior?

**Comment by willeke1234** [comment]
To synchronize my remote control app, I would like to ask GrandOrgue for the states of all MIDI objects. Currently I send MIDI events for Start MIDI recording and Stop MIDI recording, It works, but I was wondering if there's a better way.

**Comment by oleg68** [comment]
@willeke1234 so currently GO sends MIDI events only when the state of an object is changed. Do you need sending events without changing the states?

**Reply by willeke1234** [reply]
GrandOrgue does send events of every object after loading a sample set, changing settings, starting MIDI recording and other actions. I don't know if it's intentional. Last month I found out that I can rescan the MIDI devices with the Panic button. Is there a secret MIDI event to trigger the object states? Otherwise I'll use Start and Stop MIDI recording.

**Reply by oleg68** [reply]
So you need triggering to send events for all objects as a response to an incomming MIDI message, don't you?

**Reply by willeke1234** [reply]
Yes, I'll file a feature request.

**Comment by willeke1234** [comment]
>Could you share a little bit more ?

I want to use my iPad as remote control, but for now it's a macOS app. It can import an ODF and generate buttons for the visible objects on the panels. The buttons send MIDI events and change state when receiving a MIDI event. It's not very useful yet.

**Comment by midimusic** [comment]
@willeke1234 
Hi,
You appear to have the basis of a very good idea and such an app could be used to control an organ in place of any other display/control.
However you appear to be using the undocumented sysex events which are recorded at the start of every GO recording. I have spent many (happy) hours fathoming out the exact meaning of these events and have had some success.
Now I am going from memory:
Each organ (sampleset) has its own default midi channel and stop settings which are used in midi files in place of the custom settings used by your organ console. When you play a midi file these events set up the organ so it can play the file. There is a significanr snag in that the normal organ input has different default midi channels and no default stop settings.
I have cured this by feeding the organ with a settings file which makes the organ midi inputs to be the same as these default midi outputs = and this works fine in making the organ behave sensibly. Perhaps such a file could be iseful for your remote control. There are a few such settings files which you can download from my website.

It would be very useful if the organ maintainers could make GO's midi inputs default to be the same as its default outputs. This would make remote controls more practical.
csw900

**Comment by willeke1234** [comment]
As far as I'm aware, the only SYSEX event I use is the documented Hauptwerk SYSEX 32 byte LCD event. I don't use the MIDI recording file.

**Reply by midimusic** [reply]
I am not talking about the LCD events. If you think you are not using sysex why do you want to "start MIDI recording". This control is generating the undocumented sysex events I am talking about  and presumably sending them to the organ to set it up for recording midi.
If you need more convincing I suggest you look at the beginning of a GO midi recording using a midi editor.
csw900

**Reply by willeke1234** [reply]
When I start MIDI recording then GrandOrgue sends the MIDI events of all MIDI objects as set in the MIDI event dialog. In my case, Cx Program Change and Bx Controller. Maybe GrandOrgue isn't supposed to send the events but it does.

**Reply by midimusic** [reply]
GO does not send the undocumented sysex events to its output (it does not need to) but it DOES use them internally.
By the way do not forget that if you "start MIDI recording" then GO will be recording all midi events until you positively stop recording. I do not think this is a desirable outcome of your remote control.

It would be a much better idea to do as I have suggested in the past and send send these sysex events to the organ internally every time the organ is started. This would provide the correct  input settings by default and your remote control would work straight away. There would also be several other useful outcomes which I have previously suggested.
csw900

**Comment by davidgritter** [comment]
There appear to be three ways to get this information: 1: send a midi message to start recording as you suggest above, follow immediately with a stop recording message.  No sysex involved, but all the elements configured to send messages will do so to the device they've been configured to send to. 2: Reload the organ(can be configured to trigger on a midi message)  However this will set all elements to their saved value, so the current value is lost.  Plus this takes some time.  3: trigger a panic reset either with the yellow triangle on the main panel or the panic entry in the Audio/Midi menu item.  This is relatively quick, and doesn't change the state of elements, but there doesn't appear to be a way to trigger this with a midi message.  Somebody suggested in the discussion a while ago that this would be a good candidate to respond to a midi message. I would suggest the easiest way would be to put a panic button on the  Master Controls panel so that a midi message could be assigned.

**Comment by Nesselbrand67** [comment]
Are you aware that something similar does already exist from the German developer Ulrich Kirchhoff, who created the Windows program MidiStops (https://midiorgelprojekt.wordpress.com/category/midistops/)?
It is a small piece of software running on Windows and communicating with GO or HW via Midi-Messages. Configuration is done by an xml file, but an editor is also included. It also scaled perfectly with your screen size.

I used this one for controlling the Friesach sample set for my setup which is a Yamaha digital piano together with an pedalboard. I used two of the couple manuals for the piano (for the case of a split setup of the keyboard) and with the help of the 16/8/4 couplers from GO I am able even to play on two manuals (of course with some restrictions) for chorals... Also I added some controls from other panels (like the generals) and simulated LCD displays for swell and cresendo.

The only problem I have with this piece of software is due to a little bug it cannot be used for crescendo pedals (any Midi message from GO is mirrored which could lead to a loop that saturates the Midi connection).
Unfortunately, the developer seems to have stopped the work on this nice piece of software and I get no answers from him.

I am adding a screenshot so that you can get a first impression of the result. On the mentioned webpage you will find some more examples.

![Friesach](https://user-images.githubusercontent.com/113606724/190407396-deef759c-a42a-4ca5-99c4-70fd9b18f0b2.JPG)

**Reply by willeke1234** [reply]
I've seen MidiStops and I want something similar on my iPad.
There is a MIDI loop bug in GrandOrgue, I don't know if it is the same one. A workaround is to use different MIDI ports for input and output.

**Reply by midimusic** [reply]
@willeke1234 
Your "MIDI loop bug in GrandOrgue" is not a bug. It is the normal action you get when you feed output back to the input. It is called  a positive feedback loop and applies to any system, including organs.
csw900

**Reply by willeke1234** [reply]
@midimusic I don't get a positive feedback loop on buttons.

**Comment by Nesselbrand67** [comment]
I do not know about these bugs, so far GO just worked and I have not dug deeper :-).
GO is definitely mirroring incoming MIDI messages. When I use MidiStops to pull/push a stop, MidiStops sends the configured MIDI message, GO mirrors it back to MidiStops and that's it, which is fine.
But, using the crescendo, GO is the source sending first the MIDI message to MidiStops. Then, unfortunately, MidiStops also mirrors the message back to GO. Problem now is, that GO takes this incoming message as a command :-(. For normal actions no problem, but when your crescendo jitters between two steps (or you are too hectic with it...) it could happen that
1) GO pulls a stop
2) GO pushes same stop
3) MidiStops mirrors 1,3,...) due to some delay, which GO interprets as pull stop and sends pull message to MidiStops
4) MidiStops mirrors 2,4,...) due to some delay, which GO interprets as push stop and sends push message to MidiStops
-> Perfect loop, system no longer responsive, cracking in output, until I disable the virtual cable between them.
